### DISCUSSION TITLE
Allow purchase before account creation?

### DISCUSSION THREAD

**Discussion by mpvosseller** [pay-rails/pay#861] [General] [answered]
My company is interested in allowing customers to make a purchase before they create a full account with us. i.e. let them make a purchase first and then afterwards have them create an account + tie that purchase to the new account. 

The pay gem doesn't seem designed to support a use case like that (it needs a User-like object with an email address before the purchase).

Anyone know of a pattern or approach that might work?

**Comment by excid3** [comment, marked as answer]
Stripe lets you make guest purchases, so you could create a Pay::Customer afterwards if they created an account with the matching Customer ID and sync their charges.

**Reply by mpvosseller** [reply]
Got it. Thank you for the suggestion!

**Comment by mpvosseller** [comment]
@excid3 Allowing a user to make a purchase without an account (or without being logged in) opens the possibility that one real world user could end up having multiple real world stripe customers.

Do you know of any issues I might encounter (in the pay gem) if I create more than one "Stripe" `Pay::Customer` for a single user? 

i.e. I'd create a new Stripe `Pay::Customer` (with a different stripe customer id) and then pick one (maybe the most recent one) to be the default and set that with `set_payment_processor`?

Seems like I could still find the old ones using the `pay_customers` method so we wouldn't lose anything and could switch between them later if needed.

WDYT?

**Reply by excid3** [reply]
If your guest customer enters an email address, you can always look them up from the API and create the `Pay::Customer` with that Stripe ID so you don't have duplicates.

You can always cache the Stripe ID in the session so that if they do create an account within X days, you can also associate it on registration.

**Reply by mpvosseller** [reply]
Thank you. Agree with those mitigations.

In our case we use Stripe Checkout so a logged out user would enter their email address but it would be on Stripe.com and Stripe would create the "duplicate" Stripe customer.

If we get into a situation like this do you think it is safe to create and associate a second Stripe `Pay::Customer` with one user and make that one the default? i.e. Do you know of anything that might break doing that?

Thanks so much.

**Reply by excid3** [reply]
Won't cause any problems on Pay to have two `Pay::Customer`s.

But you can use their API to search for the Stripe::Customer by email so you can reuse the one they created during Checkout. Wouldn't that help you prevent creating duplicates?

**Reply by mpvosseller** [reply]
We use Stripe Checkout so one of the flows where I don't think it would be possible:

- user creates an account with us `foo@example.com`
- user starts to check out
- the pay gem creates a default stripe customer `123` for them
- user doesn't complete a purchase
- days later
- user visits our site on a different device
- they are not logged in and go through guest checkout
- we create a stripe checkout session that is not tied to any user
- user lands on stripe.com and enters their credit card and email address `foo@example.com`
- stripe automatically creates a new stripe customer `456` for this purchase
- user returns to our site and signs in
- the user is "attached" to customer `123` but we want them to be attached to customer `456`

With an "anonymous" Stripe Checkout we won't see the user's email address until after purchase and after a new Stripe customer is created.

Does that make sense? So thinking the best thing we could do is add this to the customer and maybe make it the new default.

**Reply by excid3** [reply]
You can definitely associate the Stripe Customer `456` with the User in your database that has the same email. 

If the Stripe Customer `123` doesn't have any charges or subscriptions, you could delete it then to clean up. Otherwise, you'll have two Stripe::Customer attached to a single User which is fine. If they already have charges on `123`, then I wouldn't set `456` as the default.

**Reply by mpvosseller** [reply]
Got it. Thanks so much @excid3. 🙏
